### PR TITLE
add --accounts-index-scan-results-limit-mb to allow scans to abort

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem
rpc scans can accumulate large # of results. This will take a while and time out. The scans will continue accumulating results until the scan completes.
#### Summary of Changes
Allow validators to limit the max mb of data for a scan to accumulate. Once exceeded, the scan will abort.
Fixes #

### CODE REVIEW

File: rpc/src/rpc.rs [hunk 1]
```diff
@@ -1841,11 +1841,13 @@ impl JsonRpcRequestProcessor {
                         account.owner() == program_id && filter_closure(account)
                     },
                     ScanConfig::default(),
+                    bank.mb_limit_for_scans(),
                 )
                 .map_err(|e| RpcCustomError::ScanError {
                     message: e.to_string(),
                 })?)
         } else {
+            // this path does not need to provide a mb limit because we only want to support secondary indexes
```
Comment (jeffwashington) [inline rpc/src/rpc.rs:1850]: I can plumb this through or we can ignore this case.

File: runtime/src/accounts.rs [hunk 2]
```diff
@@ -785,21 +786,54 @@ impl Accounts {
...
+                            if let Some(mb_limit) = mb_limit.as_ref() {
+                                let added = account.data().len()
+                                    + std::mem::size_of::<AccountSharedData>()
+                                    + std::mem::size_of::<Pubkey>();
+                                if sum
+                                    .fetch_add(added, Ordering::Relaxed)
+                                    .saturating_add(added)
+                                    > *mb_limit
+                                {
+                                    config.abort();
```
Comment (jeffwashington) [inline runtime/src/accounts.rs:822]: abort triggered here

File: runtime/src/accounts_index.rs [hunk 3]
```diff
@@ -104,6 +112,8 @@ pub trait IndexValue:
 pub enum ScanError {
     #[error("Node detected it replayed bad version of slot {slot:?} with id {bank_id:?}, thus the scan on said slot was aborted")]
     SlotRemoved { slot: Slot, bank_id: BankId },
+    #[error("Scan aborted")]
+    Aborted,
```
Comment (t-nelson) [inline runtime/src/accounts_index.rs:116]: It might be valuable to have this carry a "reason" field

File: runtime/src/bank.rs [hunk 4]
```diff
@@ -1520,6 +1520,14 @@ impl Bank {
         new
     }
 
+    pub fn mb_limit_for_scans(&self) -> Option<usize> {
```
Comment (t-nelson) [inline runtime/src/bank.rs:1523]: Rename?  We have bytes here, not MB

File: runtime/src/bank.rs [hunk 5]
```diff
@@ -5280,14 +5288,16 @@ impl Bank {
         &self,
         index_key: &IndexKey,
         filter: F,
-        config: ScanConfig,
+        config: &ScanConfig,
+        mb_limit: Option<usize>,
```
Comment (t-nelson) [inline runtime/src/bank.rs:5292]: rename here too

File: rpc/src/rpc.rs (hunk 1 above)
Comment (jeffwashington) [reply rpc/src/rpc.rs:1850]: punting for now

File: runtime/src/accounts.rs [hunk 6]
```diff
@@ -785,21 +786,51 @@ impl Accounts {
...
+                    });
                 },
-                config,
+                &config,
             )
-            .map(|result| result.0)
+            .map(|result| result.0);
+        if config.is_aborted() {
+            ScanResult::Err(ScanError::Aborted(
+                "The accumulated scan results exceeded the limit".to_string(),
```
Comment (t-nelson) [inline runtime/src/accounts.rs:829]: Hmm... I guess we wind up with the same abort message here regardless of whether the scan data limit was exceed or some external process flipped the flag :thinking:

File: rpc/src/rpc.rs [hunk 7]
```diff
@@ -1957,7 +1960,8 @@ impl JsonRpcRequestProcessor {
                                 }
                             })
                     },
-                    ScanConfig::default(),
+                    &ScanConfig::default(),
+                    bank.byte_limit_for_scans(),
```
Comment (t-nelson) [inline rpc/src/rpc.rs:1964]: Should this be a field in `ScanConfig` ?

File: runtime/src/accounts.rs [hunk 8]
```diff
@@ -785,21 +786,51 @@ impl Accounts {
...
         index_key: &IndexKey,
         filter: F,
-        config: ScanConfig,
+        config: &ScanConfig,
+        byte_limit: Option<usize>,
     ) -> ScanResult<Vec<(Pubkey, AccountSharedData)>> {
-        self.accounts_db
+        let sum = AtomicUsize::default();
+        let config = ScanConfig {
+            abort: Some(config.abort.as_ref().map(Arc::clone).unwrap_or_default()),
```
Comment (t-nelson) [inline runtime/src/accounts.rs:794]:
```suggestion
            abort: config.abort.as_ref().map(Arc::clone),
```

File: runtime/src/accounts.rs [hunk 9]
```diff
@@ -785,21 +786,51 @@ impl Accounts {
         bank_id: BankId,
         index_key: &IndexKey,
         filter: F,
-        config: ScanConfig,
+        config: &ScanConfig,
```
Comment (t-nelson) [inline runtime/src/accounts.rs:789]: What was the motivation for switching from move to reference here?

File: runtime/src/bank.rs [hunk 10]
```diff
@@ -5280,14 +5288,16 @@ impl Bank {
         &self,
         index_key: &IndexKey,
         filter: F,
-        config: ScanConfig,
+        config: &ScanConfig,
+        byte_limit: Option<usize>,
```
Comment (t-nelson) [inline runtime/src/bank.rs:5292]:
```suggestion
        byte_limit_for_scan: Option<usize>,
```

File: runtime/src/bank.rs [hunk 11]
```diff
@@ -5280,14 +5288,16 @@ impl Bank {
...
     ) -> ScanResult<Vec<(Pubkey, AccountSharedData)>> {
         self.rc.accounts.load_by_index_key_with_filter(
             &self.ancestors,
             self.bank_id,
             index_key,
             filter,
             config,
+            byte_limit,
```
Comment (t-nelson) [inline runtime/src/bank.rs:5300]:
```suggestion
            byte_limit_for_scan,
```

File: rpc/src/rpc.rs (hunk 7 above)
Comment (jeffwashington) [reply rpc/src/rpc.rs:1964]: I was imagining these were separate concepts. byte_limit doesn't go all the way to the scan itself. Seems more like client data.

File: runtime/src/accounts.rs (hunk 8 above)
Comment (jeffwashington) [reply runtime/src/accounts.rs:794]: I always want 'abort' to be Some if it was None. If it was Some, I want to keep that Arc. Maybe there is another method that does what I want here.

File: runtime/src/accounts.rs (hunk 9 above)
Comment (jeffwashington) [reply runtime/src/accounts.rs:789]: So I can check config.is_aborted() after calling a scan without cloning the config.

File: rpc/src/rpc.rs (hunk 7 above)
Comment (t-nelson) [reply rpc/src/rpc.rs:1964]: I see.  I suppose I saw it as a thing that effects the scan (and less plumbing).  Not a blocker either way

**Review by t-nelson — APPROVED** [review]
lgtm!

**Pull request merged**